elements/pl-parsons-problem/pl-parsons-problem.py

```diff
@@ -0,0 +1,280 @@
+        #     html += '</pre>'
+
+    else:
+        raise Exception('Invalid panel type: %s' % data['panel'])
+
+    return html
+
+
+
+def parse(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, 'answers-name')
+
+    submitted = data['submitted_answers'].get(name, '')
+    num_pieces = len(data['params'].get(name, []))
+
+    if num_pieces == 0:
+        raise Exception('number of pieces is zero')
+        
+    if len(submitted) == 0:
+        data['format_errors'][name] = 'No submitted answer.'
+        return
+
+    unpack_answer(submitted, num_pieces, data, name)        
+
+
+
+
+def grade(element_html, data):
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(element, 'answers-name')
+    weight = pl.get_integer_attrib(element, 'weight', 1)
+    check_indentation = pl.get_boolean_attrib(element, 'check-indentation', DEFAULT_CHECK_INDENTATION)
+    
+    pieces = data['params'].get(name, [])
+    correct = data['correct_answers'].get(name, [])
+    submitted_str = data['submitted_answers'].get(name, '')
+    unpacked_submitted = unpack_answer(submitted_str, len(pieces), data, name)
+
+    score, feedback, flags = grade_submitted(pieces, correct, unpacked_submitted, check_indentation)
+
+    data['partial_scores'][name] = {'score': score, 'weight': weight}
+
+
+# def test(element_html, data):
+#     element = lxml.html.fragment_fromstring(element_html)
+#     name = pl.get_string_attrib(element, 'answers-name')
+#     weight = pl.get_integer_attrib(element, 'weight', 1)
+# 
+#     correct_key = data['correct_answers'].get(name, {'key': None}).get('key', None)
+#     if correct_key is None:
+#         raise Exception('could not determine correct_key')
+#     number_answers = len(data['params'][name])
+#     all_keys = [chr(ord('a') + i) for i in range(number_answers)]
+#     incorrect_keys = list(set(all_keys) - set([correct_key]))
+# 
+#     result = random.choices(['correct', 'incorrect', 'invalid'], [5, 5, 1])[0]
+#     if result == 'correct':
+#         data['raw_submitted_answers'][name] = data['correct_answers'][name]['key']
+#         data['partial_scores'][name] = {'score': 1, 'weight': weight}
+#     elif result == 'incorrect':
+#         if len(incorrect_keys) > 0:
+#             data['raw_submitted_answers'][name] = random.choice(incorrect_keys)
+#             data['partial_scores'][name] = {'score': 0, 'weight': weight}
+#         else:
+#             # actually an invalid submission
+#             data['raw_submitted_answers'][name] = '0'
+#             data['format_errors'][name] = 'INVALID choice'
+#     elif result == 'invalid':
+#         data['raw_submitted_answers'][name] = '0'
+#         data['format_errors'][name] = 'INVALID choice'
+# 
+#         # FIXME: add more invalid choices
+#     else:
+#         raise Exception('invalid result: %s' % result)
```

Comment:
We should eventually try to implement a `test()` function for this element.  This takes a `test_type` parameter that can be one of `correct`, `incorrect`, or `invalid` and should create a test submission &mdash; you can see an example here: https://github.com/PrairieLearn/PrairieLearn/blob/master/elements/pl-number-input/pl-number-input.py#L328.